PROTOCOL.md

```diff
@@ -0,0 +1,260 @@
+
+#### 2.1.1 *CONTROL* streams
+
+Stream type : `0x00`
+
+A *CONTROL* stream is identified by the stream type `0x00`. Both the client and the server MUST open a *CONTROL* stream right after establishing the connection and MUST accept a control stream by the other peer. If either condition is not met, the connection MUST be aborted (TODO: Error code).
+
+### 2.2 Bidirectional streams
+
+The waht protocol uses only bidirectional streams opened by the client.
+
+The first byte sent on a bidirectional stream by the client defines its stream type. If a server does not support the stream type, it MUST close the stream (TODO: Error code).
+
+#### 2.2.1 *QUERY* stream
+
+Stream type : `0x01`
+
+A *QUERY* stream is opened by the client with a *QUERY* frame. The server replies with zero or more *ANNOUNCE* frames. The *ANNOUNCE* frames contain the original signatures of the announcing peers, enabling end to end verification.
+
+If the server will not send further replies, it sends an *END* frame and closes the sending side of the bidirectional stream.
+
+If the client is no longer interested in further results, it sends an *END* frame and closes both the sending and receiving side of the stream.
+
+Example:
+```
+Client -> QUERY
+Server <- ANNOUNCE
+Server <- ANNOUNCE
+Client -> END
+Server <- ANNOUNCE
+Server <- END
+Server <- no more results, close send side
+Client -> close send & recv sides
+```
+
+#### 2.2.2 *ANNOUNCE* stream
+
+Stream type : `0x02`
+
+An *ANNOUNCE* stream is opened by the client. The client sends *ANNOUNCE* frames. The server replies with *ACK* frames.
+
+Example:
+```
+Client -> ANNOUNCE
+Client -> ANNOUNCE
+Client -> ANNOUNCE
+Server <- ACK
+Server <- ACK
+Server <- ACK
+Client -> ANNOUNCE
+Client -> UNANNOUNCE 
+Server <- ACK
+Server <- ACK
+```
+
+### 3. Frames
+
+After the initial byte setting the stream type, all streams defined in section 2 carry *waht frames*. Most frames are only allowed on one or some stream types.
+
+#### 3.1 Frame layout
+
+All frames share the same layout:
+```
+Type (u32)
+Length (u32)
+Frame Payload (..)
+```
+
+### 3.2 Frame definition
+
+#### 3.2.1 `HELLO`
+
+Streams: `CONTROL` (client, server)
+
+```
+Hello {
+    PeerId (Key)
+    Capabilities (Token..)
+}
+
+Hello FRAME {
+  Type (i) = 0x01
+  Length (i)
+  Hello (Hello)
+}
+```
+
+#### 3.2.X `ANNOUNCE`
+
+Streams: `ANNOUNCE` (client), `QUERY` (server)
+
+```
+AnnouncePayload = PeerInfo | AnnounceTarget
+
+PeerInfo = {
+    Type = 0x00
+    SocketAddress (SocketAddress..)
+    Relay (RelayAddress..)
+}
+
+AnnounceTarget = {
+    Type = 0x01
+    Target (Hash)
+    Topics (Hash..)
+}
+
+Signature = {
+    Algorithm = 0x00
+    Signature (64)
+}
+
+Via = {
+    PeerId (32)
+    Signature (Signature?)
+}
+
+ANNOUNCE Frame {
+  Type (1) = 0x0X
+  Length (1)
+  PeerId (PeerId)
+  Payload (AnnouncePayload)
+  Timestamp (8)
```

Comment:
I'm worried about time-synchronisation here.  Could you replace this with "age"?  That way maybe servers only have to keep their own time but only send `Duration`s over the wire.